torchao/quantization/quantize_/workflows/int4/int4_tile_packed_to_4d_tensor.py

```diff
@@ -0,0 +1,312 @@
+        block_size: the block size for quantization, representing the granularity,
+                   for example groupwise quantization will have block_size (1, group_size)
+        shape: shape of the original Tensor
+        inner_k_tiles: An internal argument for the packing function of tile packed to 4d layout
+                      that can affect the performance of the matmul kernel. Always set to 8.
+
+    Note on Details for tile packed to 4d packing format:
+
+      This is used by tinygemm kernels `_weight_int4pack_mm`. The weight is stored as
+      a 4-d packed tensor with specific packing format for efficient computation on tensor cores.
+      The packing format optimizes for tensor core matrix multiplication performance.
+    """
+
+    tensor_data_names = ["qdata", "scale_and_zero"]
+    tensor_attribute_names = ["block_size", "shape"]
+
+    def __new__(
+        cls,
+        qdata: torch.Tensor,
+        scale_and_zero: torch.Tensor,
+        block_size: List[int],
+        shape: torch.Size,
+    ):
+        kwargs = {}
+        kwargs["device"] = qdata.device
+        kwargs["dtype"] = torch.bfloat16  # This tensor subclass only supports bfloat16
+        kwargs["requires_grad"] = False
+        return torch.Tensor._make_wrapper_subclass(cls, shape, **kwargs)  # type: ignore[attr-defined]
+
+    def __init__(
+        self,
+        qdata: torch.Tensor,
+        scale_and_zero: torch.Tensor,
+        block_size: List[int],
+        shape: torch.Size,
+    ):
+        self.qdata = qdata
+        self.scale_and_zero = scale_and_zero
+        self.block_size = block_size
+
+    def _quantization_type(self):
+        return f"shape={self.shape}, block_size={self.block_size}, device={self.device}"
+
+    @classmethod
+    def from_hp(
+        cls,
+        hp_tensor: torch.Tensor,
+        block_size: List[int],
+    ):
+        assert len(block_size) == hp_tensor.ndim, (
+            f"Expecting the length of block_size to be equal to the dimension of the weight, got {block_size=} and {hp_tensor.ndim=}"
+        )
+
+        assert all(x == 1 for x in block_size[:-1]), (
+            f"Only per group quantization is supported, got block_size: {block_size}"
+        )
+
+        assert hp_tensor.dtype == torch.bfloat16, (
+            f"Only bfloat16 is supported for Int4TilePackedTo4dTensor, got {hp_tensor.dtype}"
+        )
+
+        original_shape = hp_tensor.shape
+        # use a fixed value to simplify api
+        inner_k_tiles = 8
+
+        # Validate tensor core requirements
+        orig_out_features, orig_in_features = hp_tensor.shape[-2:]
+        # TODO: relax checks to enable quantizing in other platoforms and run in A100
+        if not torch.cuda.get_device_capability()[0] >= 8:
+            raise ValueError(
+                f"Cannot use tensor core tiled int4 kernel with a device of compute capability {torch.cuda.get_device_capability()}, the minimum compute capability is 8.0 for tensor core kernels."
```

Comment:
I don't think this subclass should say "tensor core kernels", as most kernels written for GPUs should run on tensor cores.  Is it more accurate to say "tinygemm kernels"?